sql/wsrep_mysqld.cc

```diff
@@ -2242,6 +2296,223 @@ static void wsrep_TOI_end(THD *thd) {
+    WSREP_DEBUG("%s", thd->wsrep_info);
+    thd_proc_info(thd, thd->wsrep_info);
+
+    /* DDL transaction are now atomic so append wsrep xid.
+    This will ensure transactions are logged with the given xid. */
+    bool atomic_ddl = is_atomic_ddl(thd, true);
+    if (atomic_ddl) {
+      wsrep_xid_init(thd->get_transaction()->xid_state()->get_xid(),
+                     thd->wsrep_cs().toi_meta().gtid());
+    }
+
+    ++wsrep_to_isolation;
+    rc = 0;
+  }
+
+  if (buf) my_free(buf);
+
+  thd->wsrep_gtid_event_buf_len = 0;
+  thd->wsrep_gtid_event_buf = NULL;
+
+  if (rc) wsrep_NBO_begin_failed(thd, NULL);
+
+  return rc;
+}
+
+void wsrep_nbo_isolation_begin_end(THD *thd) {
+  wsrep::mutable_buffer err;
+  WSREP_DEBUG("Ending NBO phase one");
+  int ret = thd->wsrep_cs().end_nbo_phase_one(err);
+  WSREP_DEBUG("NBO phase one ended");
+  if (ret) {
+    assert(thd->wsrep_cs().current_error());
+    WSREP_DEBUG("end_nbo_phase_one() failed for %u: %s, seqno: %lld",
+                thd->thread_id(), WSREP_QUERY(thd),
+                (long long)wsrep_thd_trx_seqno(thd));
+  }
+}
+
+int wsrep_nbo_isolation_end_begin(THD *thd) {
+  WSREP_DEBUG("NBO END maybe start");
+  wsrep::client_state &client_state(thd->wsrep_cs());
+  if (wsrep_thd_is_in_nbo(thd)) {
+    WSREP_DEBUG("NBO END start: %lld: %s",
+                client_state.toi_meta().seqno().get(), WSREP_QUERY(thd));
+    WSREP_DEBUG("NBO END start: %lld: %s",
+                client_state.transaction().ws_meta().seqno().get(),
+                WSREP_QUERY(thd));
+    // assert(strstr(WSREP_QUERY(thd), "f1") == 0);
+    int ret = client_state.begin_nbo_phase_two(thd->wsrep_nbo.keys);
+
+    if (ret) {
+      assert(client_state.current_error());
+      // HERE's the cnonconflicting issue
+      WSREP_DEBUG("begin_nbo_phase_two() failed for %u: %s, seqno: %lld",
+                  thd->thread_id(), WSREP_QUERY(thd),
+                  (long long)wsrep_thd_trx_seqno(thd));
+
+      /* jump to error handler in mysql_execute_command() */
+      switch (client_state.current_error()) {
+        default:
+          WSREP_WARN(
+              "NBO isolation failed for: %d, schema: %s, sql: %s. "
+              "Check wsrep connection state and retry the query.",
+              ret, (thd->db().str ? thd->db().str : "(null)"),
+              WSREP_QUERY(thd));
+          if (!thd->is_error()) {
+            WSREP_WARN("Adjusting error to deadlock");
+            // TODO???????
+            // my_error(ER_LOCK_DEADLOCK, MYF(0),
+            //        "WSREP replication failed. Check "
+            //       "your wsrep connection state and retry the query.");
+          }
+      }
+
+      return 1;
+    }
+  }
+  return 0;
+}
+
+static void wsrep_NBO_end(THD *thd) {
+  wsrep_to_isolation--;
+
+  wsrep::client_state &client_state(thd->wsrep_cs());
+  WSREP_DEBUG("NBO END: %lld: %s", client_state.nbo_meta().seqno().get(),
+              WSREP_QUERY(thd));
```

Comment:
Compilation fails here
```
[ 79%] Building CXX object sql/CMakeFiles/sql_main.dir/wsrep_var.cc.o
make[2]: *** [sql/CMakeFiles/sql_main.dir/build.make:141: sql/CMakeFiles/sql_main.dir/wsrep_high_priority_service.cc.o] Error 1
make[2]: *** Waiting for unfinished jobs....
In file included from /home/venki/work/pxc/review/nbo/sql/sql_class.h:165,
                 from /home/venki/work/pxc/review/nbo/sql/rpl_commit_stage_manager.h:32,
                 from /home/venki/work/pxc/review/nbo/sql/binlog.h:52,
                 from /home/venki/work/pxc/review/nbo/sql/wsrep_mysqld.cc:16:
/home/venki/work/pxc/review/nbo/sql/wsrep_mysqld.cc: In function ‘void wsrep_NBO_end(THD*)’:
/home/venki/work/pxc/review/nbo/sql/wsrep_mysqld.cc:2484:49: error: ‘class wsrep::client_state’ has no member named ‘nbo_meta’; did you mean ‘wsrep::ws_meta wsrep::client_state::nbo_meta_’? (not accessible from this context)
 2484 |   WSREP_DEBUG("NBO END: %lld: %s", client_state.nbo_meta().seqno().get(),
      |                                                 ^~~~~~~~
/home/venki/work/pxc/review/nbo/sql/wsrep_mysqld.h:283:25: note: in definition of macro ‘WSREP_LOG’
  283 |         .message(fmt, ##__VA_ARGS__);                              \
      |                         ^~~~~~~~~~~
/home/venki/work/pxc/review/nbo/sql/wsrep_mysqld.cc:2484:3: note: in expansion of macro ‘WSREP_DEBUG’
 2484 |   WSREP_DEBUG("NBO END: %lld: %s", client_state.nbo_meta().seqno().get(),
      |   ^~~~~~~~~~~
```
Can you check this?

Reply:
https://github.com/percona/wsrep-lib/pull/7 

Looks like I didn't add this function in my wsrep-lib change PR, but I had it in my local NBO checkout.

Reply:
Can you please merge the wsrep-lib PR and update this branch?

Reply:
done.